content/en/docs/concepts/extend-kubernetes/compute-storage-net/resource-drivers.md

```diff
@@ -0,0 +1,230 @@
+to the resource driver, which manages such resources in the cluster.
+
+A resource driver consists of two main components:
+
+- a _controller_ (one per cluster), manages hardware resources allocation for
+  {{< glossary_tooltip term_id="resource-claim" text="ResourceClaims">}}
+- _kubelet plugin_ (one per node that has or can access the associated resource), that:
+  - discovers the supported hardware
+  - announces the discovered hardware to the resource driver controller
+  - prepares the hardware allocated to a ResourceClaim when the Kubelet prepares to create the Pod
+  - unprepares the hardware allocated for a Pod when the Pod has reached final state or is being deleted.
+
+There are two common ways to implement communication between the controller and a kubelet plugin:
+
+- through custom resource objects that use
+  {{< glossary_tooltip term_id="CustomResourceDefinition" text="CustomResourceDefinitions">}}
+ provided by the vendor or project behind the resource driver
+- through a ResourceHandle which is a part of an `AllocationResult` provided by the controller in case of
+  successful allocation
+
+General recommendations:
+
+- resource driver name pattern: `<HW type>.resource.<companyname>.<companydomain>`. For example,
+  gpu.resource.example.com
+
+<!-- body -->
+
+## Implementing controller
+
+Communication between the scheduler and a resource driver is done through
+{{< glossary_tooltip term_id="pod-scheduling-context" text="PodSchedulingContext">}} API object. It is
+recommended to use [controller helper library](/docs/reference/helper-libraries/dra-driver-controller/),
+it implements all needed operations related to PodSchedulingContext, they are common for all resource
+drivers.
+
+When using the DRA helper code, resource driver controller has to implement a
+[driver interface](https://pkg.go.dev/k8s.io/dynamic-resource-allocation@v0.28.1/controller#Driver),
+which then can be simply used with
+[controller helper instance](https://pkg.go.dev/k8s.io/dynamic-resource-allocation@v0.28.1/controller#New).
+
+See the [example resource driver controller](https://github.com/kubernetes-sigs/dra-example-driver/blob/151b7c8da2e620c47a3591e1a937f8d0297b0c25/cmd/dra-example-controller/main.go#L208) for details.
+
+A resource driver controller's main responsibility is to allocate and deallocate resources for
+{{< glossary_tooltip term_id="resource-claim" text="ResourceClaims">}}.
+
+There are two modes of allocation the ResourceClaim can have:
+
+- `WaitForFirstConsumer` (default), which you could think of as meaning _delayed_.
+  In this mode the cluster only requests resource(s)
+  for ResourceClaim when a Pod that needs it is being scheduled.
+- `Immediate`: the resource has to be allocated to the ResourceClaim as soon as possible, and
+  retained until the ResourceClaim is deleted.
+
+If `allocationMode` is not set explicitly, the mode is `WaitForFirstConsumer`.
```

Comment:
I'd move this level of detail into a separate page within https://kubernetes.io/docs/reference/node/, and write a shorter text about why PodSchedulingContext and ResourceClaim are used.

Reply:
While I agree that reference spec of the ResourceClaim object is better placed in `reference/node`, this is supposed to be an article about extending Kubernetes, how to write a resource driver. Implementing a controller part of a resource driver seems to me very much suitable here. We don't bring in the whole info about the ResourceClaim or PodSchedulingContext in here, this is in fact a short summary, only what's important to keep in mind when writing a resource driver.

Reply:
Okay, so the "extending Kubernetes" is not for writing extensions, but about customizing the instance of Kubernetes, like installation of device plugins and CRDs, and in our case resource drivers. got it.

Reply:
Minimized the chapter.